### ISSUE TITLE
How to build and package modified sources

### ISSUE THREAD

**Issue by castler** [conan-io/conan#1160]
I'm sorry but I don't really get how the workflow with conan would look like in the following use case:

1. Checkout sources into user workspace (not possible with conan right? since source() would checkout the sources into the local cache)
2. Modify sources in the user workspace
3. Build libs from modified sources (either with plain cmake or with conan?)
4. Package the libs and header files from the previous build and store in local conan cache

Thanks for your help

**Comment by memsharded** [comment]
``conan source`` can also execute locally the ``source()`` method in user space. Then you could probably use the ``conan build`` and ``conan package`` commands, also locally to do the whole process locally. We are working in a process to put binaries and artifacts in general from  user space into conan packages in the local cache.

However, the workflow that you expose has a big problem is that it is not reproducible. If you modify the sources locally, then create the lib and package the libs, you cannot guarantee that those libs can be rebuilt, because it is not in sync with the source code origin. Conan puts strong focus on reproducibility of the package creation.

If you are concerned about how to work in your packages locally, maybe there are other flows that you might consider: One possibility is to work locally, from the sources repo, with the sources embedded in the recipe with the ``exports_sources`` just as a consumer. The flow could be like:

```
- git clone your sources (can be a package)
- conan install (so your package gets also its depedencies installed)
- cmake -G "Visual "  #use your standard build, open your IDE, build, debug...
- conan test_package  # when you are satisfied with your local builds and want to test the full package creation. 
```

``test_package`` is the strongly recommended way to create conan packages.

It is true that the last step will re-build the package inside the local conan cache, but as suggested, this is a very recommended step. In fact, the best would be to also commit the changes, push, and let the CI build the package for different configurations. As said before, we are working on a new conan command that will allow to copy your local binaries to a conan package, but it will be at the expense of reproducibility guarantees.

Please tell me if this helps or you have further questions.

**Comment by castler** [comment]
The new feature you were talking about sounds awesome, is there already an existing issue?

You are totally right that those packages build from locally modified sources won't be able to be reproduced. But imagine there are two closely coupled packages. In order to fix a bug the developer has to change the source of both packages. Before he commits his changes he wants to build both packages locally and see if the bug is gone. In this case reproducible is not important for me, since in the next step the developer commits his changes and the CI builds the reproducable packages. 

I read about the "exports_sources" option. I did not use it, because if I understood it correctly, this will duplicate the source code on my machine, but you're right this may be a good solution  when it comes to  reproducability also with local builds. I will try it :)

Yes your answer helped me a lot, Thank you!

**Comment by memsharded** [comment]
Yes, there is already a PR :)
Check https://github.com/conan-io/conan/pull/1152

About the ``exports_sources``, the fact is that the source code will always be "duplicated" in the "source" folder of the  local cache. If the source code is embeded in the recipe with ``exports_sources`` it will be downloaded (only when necessary to build from sources) from the conan remote as a tgz, then unzipped, and if using the ``source()`` method, it will typically execute a "git clone" or a download to get the code from a external origin into the source folder. So overall, the thing is that there are two different processes, one for actively developing code, as a consumer (in user folders), and one for creating packages (in the conan local cache). You probably want to check also this issue: https://github.com/conan-io/conan/issues/1001

**Comment by memsharded** [comment]
Hi!

The new ``package_files`` feature, has been released in conan 0.22.
You can check the docs for the new feature: http://docs.conan.io/en/latest/howtos/existing_binaries.html
and here: http://docs.conan.io/en/latest/reference/commands/package_files.html

Also, there is already an open issue for next 0.23 to review the development of packages: https://github.com/conan-io/conan/issues/1171. With that covered, I think this current question has been addressed and can be closed, but please ask any further question or re-open if necessary. Thanks!
